# First, let's take a look at the Basel Problem


### What's the Basel problem I hear you ask?

Well imagine all the numbers up to ... well infinity?

Square them, and then inverse them.


$1+\frac{1}{2^2} + \frac{1}{3^2} + \frac{1}{4^2} + \frac{1}{5^2} + ... $ 

Well, imagine if the sum of all that, all the way to infinity (and naturally beyond if Buzz is able to find that) was:

$\pi^2 / 6$

#### Mad? 
#### Doesn't even make sense?
#### Going to stop reading?

Well when dear Leo Euler showed it to the world, the world thought very much the same. Fortunately for him, us and modern society, they kept reading...


Annoyingly to understand the Basel Series, you'll need to understand something called the Taylor Series. Which is a lot less fun, but also important if you like triangles.

In [1]:
from math import pi,sin,factorial

#### Now remember that sine is the ratio of the length of the side that is opposite that angle to the length of the longest side of the triangle (the hypotenuse)

https://www.google.com/url?sa=i&source=images&cd=&ved=2ahUKEwia1tOniMfkAhWByoUKHYYEBTQQjRx6BAgBEAQ&url=https%3A%2F%2Fen.wikipedia.org%2Fwiki%2FSine&psig=AOvVaw0W8HTgs6j3PsiSifF4V7ZS&ust=1568232728636529


## As a result, it's linked to a π via the circle connotation, but if you want to get a decent visulisation it's...

In [2]:
import numpy as np

import matplotlib.pyplot as plt

 

# Get x values of the sine wave
time = np.arange(0, 10, 0.1);

 

# Amplitude of the sine wave is sine of a variable like time
amplitude = np.sin(time)

# Plot a sine wave using time and amplitude obtained for the sine wave
plt.plot(time, amplitude)

 

# Give a title for the sine wave plot
plt.title('Sine wave')

 

# Give x and y axis labels for the sine wave plot
plt.xlabel('Time')
plt.ylabel('Amplitude = sin(time)')

plt.grid(True, which='both')

plt.axhline(y=0, color='k')

### So from above you can see that the graph is periodic (swings back and forth... amirite?) and that it alternates between 1 and -1.

But what about those times where x = 0?

In [3]:
for i in np.linspace(0,4,20):
    print(i,"->",sin(i))


0.0 -> 0.0
0.21052631578947367 -> 0.20897462406278547
0.42105263157894735 -> 0.40872137322898616
0.631578947368421 -> 0.5904198559291864
0.8421052631578947 -> 0.7460466536513234
1.0526315789473684 -> 0.8687296180358695
1.263157894736842 -> 0.9530513303299016
1.4736842105263157 -> 0.995288323076044
1.6842105263157894 -> 0.9935755005153384
1.894736842105263 -> 0.9479884972487855
2.1052631578947367 -> 0.860540338374244
2.3157894736842106 -> 0.7350925485799386
2.526315789473684 -> 0.5771846354279316
2.7368421052631575 -> 0.3937894764830594
2.9473684210526314 -> 0.193005411871413
3.1578947368421053 -> -0.016301361193955085
3.3684210526315788 -> -0.22488830082300723
3.578947368421052 -> -0.42354465143912384
3.789473684210526 -> -0.6034981711535158
4.0 -> -0.7568024953079282


#### You can see it's between 2.9473684210526314 and 3.1578947368421053.

But let's do...

In [4]:
for i in np.linspace(3,3.2,10):
    print(i,"->",sin(i))


3.0 -> 0.1411200080598672
3.022222222222222 -> 0.1190871424307064
3.0444444444444443 -> 0.0969954707562093
3.066666666666667 -> 0.0748559020549272
3.088888888888889 -> 0.052679368997292045
3.111111111111111 -> 0.030476822506986412
3.1333333333333333 -> 0.008259226353304283
3.155555555555556 -> -0.013962448262834113
3.177777777777778 -> -0.03617722812653456
3.2 -> -0.058374143427580086


Well, we can get it bang on surely?

In [5]:
0 == round(sin(pi),15)

True

Just don't do...

In [6]:
0 == round(sin(pi),16)

False

### But anyway, because of how right angle triangles relate to circles, when $sin(x)$ == 0, it can be π, 2π, 3π or even -π...

The point that matters is that is how the sine ratio works with 0. 0 is terribly important if you want to solve functions as it's a great indicator of what the other coefficients are when everything else is minimised.


Now that rounding to 15 or 16 decimal places is because there's a slight discrepancy of the number of figures the sin value from the `math` library is I'm using. So what is $sin(x)$ really?

$sin(x) = x -\frac{x^3}{3!} + \frac{x^5}{5!} + \frac{x^7}{7!} + \frac{x^9}{9!} + ...  = 0$ 


That `!` is the same a running our `factorial` function from Chapter 2.

Now in the `math.sin` function, that calculation above only goes to 15 places. 


But what this also means is:

$sin(x) = (x-0)(x-\pi)(x+\pi)(x-2\pi)(x+2π)(x-3\pi)...$

Because for every iteration of π, `sin(nπ)` (with n being a whole number) will be 0.


Now what this will also mean is because:

$(x-\pi)(x+\pi) == (x^2 - \pi^2)$

$(x-2\pi)(x+2\pi) == (x^2 - 4\pi^2)$

$(x-3\pi)(x+3\pi) == (x^2-9\pi^2)$

(...hold on that's where those that's where those square numbers are coming in...)


### It also means that:

$sin(x) = (x^2-\pi^2)(x^2-4\pi^2)(x^2-9\pi^2)...$


What if we just divide the right side of that equation by $x^2$ within each bracket?


$sin(x) = (1 - \frac{x^2}{π^2})(1 - \frac{x^2}{4\pi^2})(1 - \frac{x^2}{9\pi^2})...$


Then divide both sides by `x`...


$\frac{sin(x)}{x} = (1 - \frac{x^2}{\pi^2})(1 - \frac{x^2}{4\pi^2})(1 - \frac{x^2}{9\pi^2})...$


#### And break that down a little further... 

$\frac{sin(x)}{x} = 1 - \frac{x^2}{\pi^2}  - \frac{x^2}{4\pi^2} - \frac{x^2}{9\pi^2}...$

$\frac{sin(x)}{x} = 1 - x^2 (\frac{1}{\pi^2}  - \frac{1}{4\pi^2} - \frac{1}{9\pi^2}...)$

### OK still with me? It means let's take that first calculation for `sin` from the top:

$sin(x) = x -\frac{x^3}{3!} + \frac{x^5}{5!} + \frac{x^7}{7!} + \frac{x^9}{9!} + ...  = 0$ 

Divide the whole thing by `x`:

$\frac{sin(x)}{x} = 1 -\frac{x^2}{3!} + \frac{x^4}{5!} + \frac{x^6}{7!} + \frac{x^8}{9!} + ...  = 0$ 


It also means...

$1 -\frac{x^2}{3!} + \frac{x^4}{5!} + \frac{x^6}{7!} + \frac{x^8}{9!} + ...  = 1 - x^2 (\frac{1}{\pi^2}  - \frac{1}{4\pi^2} - \frac{1}{9\pi^2}...)$


### Now... (almost there I promise), you see that both sides have an $x^2$ immediately after the `1 - ...` it means that the coefficients must be the same for both functions.

In [7]:
factorial(3)

6

It means


$ 0-\frac{1}{6} = (\frac{1}{\pi^2}  - \frac{1}{4\pi^2} - \frac{1}{9\pi^2}...)$

Which is 


$\frac{1}{6} = \frac{1}{\pi^2}(1+\frac{1}{4}+\frac{1}{9}...)$


# WHICH MEANS

$\frac{\pi}{6} = 1+\frac{1}{4}+\frac{1}{9}...$

and...

$\frac{\pi}{6} = 1+\frac{1}{2^2}+\frac{1}{3^2}+\frac{1}{4^2}+\frac{1}{5^2}...$

# BOOM!!!




lovely...




## OK secondly... (oh yes we have a lot more to do), there's some basic axioms we've got to get down.

$x^m  \times x^n = x^{m+n} $

Should be pretty simple




In [8]:
12**3

1728

In [9]:
12**5

248832

In [10]:
12**3 * 12**5

429981696

In [11]:
12**8 == 12**3 * 12**5

True

As a consequence, this also means 

$x^m \div x^n = x^{m-n} $

$(x^m)^n = x^{m  \times  n} $

You can test that out yourself.


There's also:

$x^0 => 1$ Any positive number to the power of `0` will be `1`


In [15]:
np.random.randint()

TypeError: randint() takes at least 1 positional argument (0 given)